### PR TITLE
Move more json instances

### PR DESCRIPTION
**Pull request by lehins**
# Description

Moves some more instance from cardano-cli.

Fixes Arbitrary instances that got broken in #3285

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [x] Any changes are noted in the `CHANGELOG.md` for affected package
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Arbitrary.hs [hunk 1]
```diff
@@ -163,7 +163,7 @@ instance
   arbitrary =
     AlonzoTxOut
       <$> arbitrary
-      <*> arbitrary
+      <*> scale (`div` 15) arbitrary
```
Comment (teodanciu) [inline eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Arbitrary.hs:166]: I see this `scale (div 15) ` occurs in our code a lot, so I should know what it is, but I don't :grimacing: 
Could you explain where the 15 is coming from? ( sorry)

File: eras/mary/impl/src/Cardano/Ledger/Mary/Value.hs [hunk 2]
```diff
@@ -135,23 +135,36 @@ instance DecCBOR AssetName where
...
+    , ToJSONKey
+    , FromJSONKey
+    )
 
 -- | The MultiAssets map
 newtype MultiAsset c = MultiAsset (Map (PolicyID c) (Map AssetName Integer))
   deriving (Show, Generic, ToJSON)
 
 instance Crypto c => Eq (MultiAsset c) where
-  (MultiAsset x) == (MultiAsset y) = pointWise (pointWise (==)) x y
```
Comment (teodanciu) [inline eras/mary/impl/src/Cardano/Ledger/Mary/Value.hs:145]: haha almost sure these are my parentheses!  It's not possible to use them in a more useless way :laughing:

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Arbitrary.hs (hunk 1 above)
Comment (lehins) [reply eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Arbitrary.hs:166]: From my :brain: 
JK. :smile: 

Are you familiar with the concept of size in QuickCheck generators? It allows you to change the magnitude of the type you are trying to generate. For example, here are default values generated for a list of `Int`:


```haskell
$ ghci -package QuickCheck
GHCi, version 9.2.6: https://www.haskell.org/ghc/  :? for help
ghci> import Test.QuickCheck
ghci> generate (listOf arbitrary :: Gen [Int])
[28,-7,15,-25,-17,20,1,-30,-2,-3,2,11,8,-23,23,-24,28,-22,-4,17,-4,3,-29,-5,2,17,9,11]
```

You can see that values for the `Int` are pretty small. It is desired sometimes, but actually more often than not you want to increase their magnitude you can do so by, scaling up the size:

```haskell
ghci> generate (listOf (scale (* 20) arbitrary) :: Gen [Int])
[535,427,133,174,598,-338,-60,76,370,-269,94,-69,293,-507,-46,-489,-181,343,-260,-585,178,-51,330,-191,-449,-470,567,-301,418,558]
```

You can also scale up the list generator as well as the  `Int` genrator:
```haskell
ghci> generate (scale (* 20) (listOf arbitrary) :: Gen [Int])
[-237,483,-495,-550,-592,187,448,-339,355,-403,239,451,-471,-68,-550,399,-298,461,36,-340,-371,468,160,-35,-376,197,544,332,-184,286,-248,-384,-552,-315,-108,-321,512,-302,-25,-402,-362,-481,284,-276,-163,490,-338,-238,116,-584,-9,-486,-427,-399,195,-293,11,-510,-574,-578,-101,169,-462,-304,-567,-130,-337,26,569,-468,235,262,248,-573,-34,-421,-400,84,228,-82,1,-104,448,549,64,-458,-5,-279,415,-202,318,-409,545,180,-454,422,-502,542,-473,266,496,85,-563,5,-487,374,-63,-590,562,-461,451,-166,263,-549,-431,507,-272,-163,571,283,-230,-152,235,-224,228,-110,-151,-65,108,-265,372,-554,-347,221,393,-521,-566,-79,259,-494,-130,294,540,412,239,271,-111,341,567,-514,345,-510,-74,-223,-318,-524,272,317,-377,-484,-81,-151,13,-250,24,328,-327,275,-20,-491,-25,363,219,-306,-341,-45,-197,533,-382,-466,462,103,66,89,-467,557,407,-361,-513,56,-576,501,256,-532,349,-502,294,-30,541,61,542,154,-586,-303,262,-165,-283,157,-51,-349,-183,461,318,490,240,-253,572,-439,-554,568,-62,-488,-546,582,-148,-76,288,332,-82,188,-470,155,446,-296,472,536,287,-347,-177,-66,-247,54,337,-548,-301,78,449,-486,-567,138,235,-515,107,-598,-513,-283,-447,-231,571,-238,-38,-239,-325,-146,-552,412,273,192,309,-424,575,386,-76,195,-143,53,178,332,-151,-188,-285,-170,321,-462,178,-467,192,428,359,339,-448,231,-13,36,-504,-10,241,-505,-202,279,178,-111,-498,28,323,332,-192,-469,-80,-10,-270,23,-591,427,-357,193,-542,130,-192,278,125,281,161,-16,-485,107,211,-135,167,88,-442,0,-442,-250,-540,492,-390,310,-498,19,401,-457,-407,246,-163,432,99,-527,28,-433,512,-179,-465,311,-470,280,-488,-587,72,448,130,-361,-192,-36,-512,592,-204,7,-363,-492,-573,295,396,-187,-276,-477,551,328,131,392,-248,-268,487,544,541,136,-378,322,-38,449,-477,-122,222,596,-513,-278,46,-423,-126,229,-11,-580,-358,137,365,-388,0,-383,18,-488,558,8,-547,143,-15,-62]
```

In the same way that we can scale up generators, we can also scale them down:
```haskell
ghci> generate (scale (`div` 15) (listOf arbitrary) :: Gen [Int])
[1,2]
```

The default value for `size` is `30`:
```
ghci> generate (scale (error . show) arbitrary :: Gen Int)
*** Exception: 30
CallStack (from HasCallStack):
  error, called at <interactive>:18:18 in interactive:Ghci12
```

So, the number `15` was decided on empirically by looking at MultiAssets being generated

File: eras/mary/impl/src/Cardano/Ledger/Mary/Value.hs [hunk 3]
```diff
@@ -135,23 +135,36 @@ instance DecCBOR AssetName where
...
+  deriving
+    ( Show
+    , Eq
+    , Ord
+    , Generic
+    , NoThunks
+    , NFData
+    , EncCBOR
+    , DecCBOR
+    , ToJSON
```
Comment (teodanciu) [inline eras/mary/impl/src/Cardano/Ledger/Mary/Value.hs:147]: Are the default implementations of `ToJSON` and `ToJSONKey`  the same as [the one we removed](https://github.com/input-output-hk/cardano-ledger/pull/3338/commits/d1ca75f073cb84b7eab70dd83b4522dc3df71b75#diff-abef920a14ec558ef5e82f62b48e09747f4683b5f60ae07e370fc2e6b70c2f64L401)? 
```
toJSON (PolicyID (ScriptHash h)) = Aeson.String (Hash.hashToTextAsHex h)

toJSONKey = toJSONKeyText render
    where
      render (PolicyID (ScriptHash h)) = Hash.hashToTextAsHex 
```

Comment (lehins) [reply eras/mary/impl/src/Cardano/Ledger/Mary/Value.hs:147]: Good question and yes they are. I confirmed it before switching to the derived method:
https://github.com/input-output-hk/cardano-base/blob/dd15fa63bafad152873f6f257b3fed6ef37ebe78/cardano-crypto-class/src/Cardano/Crypto/Hash/Class.hs#L304-L312

File: eras/mary/impl/src/Cardano/Ledger/Mary/Value.hs [hunk 4]
```diff
@@ -382,11 +389,11 @@ decodeIntegerBounded64 = do
 -- ========================================================================
 -- JSON
 
-instance ToJSON (MaryValue era) where
+instance Crypto c => ToJSON (MaryValue c) where
   toJSON = object . toMaryValuePairs
   toEncoding = Aeson.pairs . mconcat . toMaryValuePairs
 
-toMaryValuePairs :: Aeson.KeyValue a => MaryValue crypto -> [a]
+toMaryValuePairs :: Crypto c => Aeson.KeyValue a => MaryValue c -> [a]
```
Comment (teodanciu) [inline eras/mary/impl/src/Cardano/Ledger/Mary/Value.hs:396]: Just curious, is it the change of the PolicyId `ToJSON` serializer that made it necessary to add the `Crypto c` constraint?

Comment (lehins) [reply eras/mary/impl/src/Cardano/Ledger/Mary/Value.hs:396]: It is required for `ScriptHash` deriving, which is also added in this PR:
```haskell
deriving newtype instance Crypto c => ToJSONKey (ScriptHash c)

deriving newtype instance Crypto c => FromJSONKey (ScriptHash c)
```

Comment (lehins) [reply eras/mary/impl/src/Cardano/Ledger/Mary/Value.hs:396]: So yeah, change to the `PolicyID` was the cause

**Review by teodanciu — APPROVED** [review]
Looks good to me, thank you for the explanations!

**Pull request merged**